tls/s2n_client_cert_verify.c

```diff
@@ -73,20 +76,21 @@ int s2n_client_cert_verify_send(struct s2n_connection *conn)
     POSIX_GUARD(s2n_handshake_get_hash_state(conn, chosen_sig_scheme.hash_alg, &hash_state));
     POSIX_GUARD(s2n_hash_copy(&conn->handshake.ccv_hash_copy, &hash_state));
 
-    struct s2n_cert_chain_and_key *cert_chain_and_key = conn->handshake_params.our_chain_and_key;
+    S2N_ASYNC_PKEY_SIGN(conn, conn->secure.client_cert_sig_scheme.sig_alg, &conn->handshake.ccv_hash_copy, s2n_client_cert_verify_send_complete);
 
-    DEFER_CLEANUP(struct s2n_blob signature = {0}, s2n_free);
-    uint32_t max_signature_size = 0;
-    POSIX_GUARD_RESULT(s2n_pkey_size(cert_chain_and_key->private_key, &max_signature_size));
-    POSIX_GUARD(s2n_alloc(&signature, max_signature_size));
+    return S2N_SUCCESS;
+}
 
-    POSIX_GUARD(s2n_pkey_sign(cert_chain_and_key->private_key, chosen_sig_scheme.sig_alg, &conn->handshake.ccv_hash_copy, &signature));
+int s2n_client_cert_verify_send_complete(struct s2n_connection *conn, struct s2n_blob *signature)
+{
+    struct s2n_stuffer *out = &conn->handshake.io;
 
-    POSIX_GUARD(s2n_stuffer_write_uint16(out, signature.size));
-    POSIX_GUARD(s2n_stuffer_write(out, &signature));
+    POSIX_GUARD(s2n_stuffer_write_uint16(out, signature->size));
+    POSIX_GUARD(s2n_stuffer_write(out, signature));
+    POSIX_GUARD(s2n_free(signature));
 
```

Comment:
This is not the right place to do this free-- what if `s2n_client_cert_verify_send_complete` is never called, potentially because something failed? It also looks like the server cert path doesn't do this: https://github.com/aws/s2n-tls/blob/main/tls/s2n_tls13_certificate_verify.c#L93-L101

Is the async pkey logic supposed to handle the free?

Reply:
Yes you are right, the pkey logic handles the free, I must have interpreted the valgrind results incorrectly (Locally it wasn't leaking before this change).

I think the real issue was using `conn->secure.client_cert_sig_scheme.sig_alg` instead of `chosen_sig_scheme.sig_alg` when calling sign.